File: call-in-script/ch-origin.pl
#!/usr/bin/env perl
use strict;
use warnings;

my ($year, $month, $day, $hour, $min, $sec, $msec)
    = (1981, 3, 28, 10, 23, 32, 456);

my $doy = &ymd2doy($year, $month, $day);

open(SAC, "| sac");
print SAC "echo on\n";
print SAC "r seis.SAC \n";
print SAC "ch o gmt $year $doy $hour $min $sec $msec\n";
print SAC "ch allt (0 - &1,o&) iztype IO \n";
print SAC "wh \n";
print SAC "q\n";
close(SAC);

sub ymd2doy {
    my ($year, $month, $day) = @_;
    my @noleap = qw ( 0 31 59 90 120 151 181 212 243 273 304 334 );
    my @leap   = qw ( 0 31 60 91 121 152 182 213 244 274 305 335 );

    if (($year%4 == 0 && $year%100 != 0) || ($year%400==0)){
        return $leap[$month-1]+$day;
    } else {
        return $noleap[$month-1]+$day;
    }
}

File: call-in-script/ch-origin.py
#!/usr/bin/env perl
# -*- coding: utf-8 -*-
import os
import datetime
import subprocess

year = 1981
month = 3
day = 28
hour = 10
min = 23
sec = 32
msec = 456

# arguments: year, month, day, hour, minute, second, microsecond
event = datetime.datetime(year, month, day, hour, min, sec, msec*1000)
# day of year
doy = int(event.strftime('%j'))

os.putenv("SAC_DISPLAY_COPYRIGHT", "0")

p = subprocess.Popen(['sac'], stdin=subprocess.PIPE)

s = ""
s += "r *.SAC\n"
s += "synchronize \n"
s += "ch o gmt %d %d %d %d %d %d\n" % (year, doy, hour, min, sec, msec)
s += "ch allt (0 - &1,o&) iztype IO\n"
s += "wh\n"
s += "q\n"
p.communicate(s.encode())
